idx/cassandra/cassandra.go

```diff
@@ -546,55 +560,139 @@ func (c *CasIdx) ArchiveDefs(defs []schema.MetricDefinition) (int, error) {
 }
 
 func (c *CasIdx) processWriteQueue() {
+	defer c.wg.Done()
+
 	var success bool
 	var attempts int
 	var err error
 	var req writeReq
 	qry := fmt.Sprintf("INSERT INTO %s (id, orgid, partition, name, interval, unit, mtype, tags, lastupdate) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)", c.Config.Table)
-	for req = range c.writeQueue {
-		if err != nil {
-			log.Errorf("Failed to marshal metricDef: %s. value was: %+v", err, *req.def)
-			continue
-		}
-		statQueryInsertWaitDuration.Value(time.Since(req.recvTime))
-		pre := time.Now()
-		success = false
-		attempts = 0
-
-		for !success {
-			if err := c.Session.Query(
-				qry,
-				req.def.Id.String(),
-				req.def.OrgId,
-				req.def.Partition,
-				req.def.Name,
-				req.def.Interval,
-				req.def.Unit,
-				req.def.Mtype,
-				req.def.Tags,
-				req.def.LastUpdate).Exec(); err != nil {
-
-				statQueryInsertFail.Inc()
-				errmetrics.Inc(err)
-				if (attempts % 20) == 0 {
-					log.Warnf("cassandra-idx: Failed to write def to cassandra. it will be retried. %s. the value was: %+v", err, *req.def)
+	for {
+		select {
+		case req = <-c.writeQueue:
+			if err != nil {
+				log.Errorf("cassandra-idx: failed to marshal metricDef: %s. value was: %+v", err, *req.def)
+				continue
+			}
+			statQueryInsertWaitDuration.Value(time.Since(req.recvTime))
+			pre := time.Now()
+			success = false
+			attempts = 0
+
+			for !success {
+				select {
+				case <-c.shutdown:
+					log.Info("cassandra-idx: received shutdown, exiting processWriteQueue")
+					return
+				default:
+					c.sessionLock.RLock()
+					if err := c.Session.Query(
+						qry,
+						req.def.Id.String(),
+						req.def.OrgId,
+						req.def.Partition,
+						req.def.Name,
+						req.def.Interval,
+						req.def.Unit,
+						req.def.Mtype,
+						req.def.Tags,
+						req.def.LastUpdate).Exec(); err != nil {
+
+						statQueryInsertFail.Inc()
+						errmetrics.Inc(err)
+						if (attempts % 20) == 0 {
+							log.Warnf("cassandra-idx: Failed to write def to cassandra. it will be retried. %s. the value was: %+v", err, *req.def)
+						}
+						sleepTime := 100 * attempts
+						if sleepTime > 2000 {
+							sleepTime = 2000
+						}
+						time.Sleep(time.Duration(sleepTime) * time.Millisecond)
+						attempts++
+					} else {
+						success = true
+						statQueryInsertExecDuration.Value(time.Since(pre))
+						statQueryInsertOk.Inc()
+						log.Debugf("cassandra-idx: metricDef %s saved to cassandra", req.def.Id)
+					}
+					c.sessionLock.RUnlock()
 				}
-				sleepTime := 100 * attempts
-				if sleepTime > 2000 {
-					sleepTime = 2000
+			}
+		case <-c.shutdown:
+			log.Info("cassandra-idx: received shutdown, exiting processWriteQueue")
+			return
+		}
+	}
+}
+
+func (c *CasIdx) deadConnectionCheck() {
+	defer c.wg.Done()
+
+	ticker := time.NewTicker(time.Second * 5)
```

Comment:
i'd make that configurable instead of hard coding `5`. even though most users will probably never need to tune that, it still seems better to make it configurable than to hard code it.

Reply:
Yeah I did plan on making it configurable, just wanted to test it and make sure it worked first